### ISSUE TITLE
iOS - SSO not working between mobile App and Safari (default) Browser

### ISSUE THREAD

**Issue by chookhen60** [auth0/react-native-auth0#462]
### Describe the problem
No matter what we seem to do, once a user performs a successful log in to our Mobile App in iOS, SSO does not seem to be achieved with the Safari (default) browser.

Note that all the testing I have performed thus far is with the iOS Simulator simulating an iPhone 13 running iOS 15.0.

In our case, we are using Microsoft B2C as the IdP.  We can see from examining the network traffic that a persistent Cookie is returned at the end of the login process which according to the `AsWebAuthenticationSession` documentation, should then be made available to the external Safari Browser for SSO purposes.

We have a website which uses the same B2C login/IdP mechanism, and expect that once having manually logged into the App, the user should be able to go to Safari and access this website without needing to log in.  This does not happen however, and the user is forced to log in again.

We have this working perfectly in Android (between app and website accessed via Chrome (default browser), so am confident any B2C configuration is correct.  It also works perfectly if the mobile login pathway is replicated (via using the b2c login URL exactly as is presented in the App) in the Safari browser on iOS, and then have the user browse to the website (i.e, SSO is achieved).

From examining the web traffic, it seems that Safari just does not have the persistent Cookie present/available, and because of this reason, the user is forced to login again.

### What was the expected behavior?
Once the user manually logs into the App, they are then able to go to the browser and access the Website via SSO without needing to log in again.

### Reproduction
 - Log into a Mobile App on iOS which is configured to support SSO through a persistent Cookie
 - Once logged in, using Safari (assuming this is the default browser) go to a website which uses the same IdP and attempt to login - it should perform SSO, but is instead asking the user to log in

### Environment

- library version: 2.4.0
- RN 0.60.6
- `ephemeralSession` is set to false for the login

Firstly, I really wanted to confirm that this should be working as I expect, and it would be good to hear from anyone who may have something similar set up and working, so I at least know it is possible.  

I am not sure where else I can really perform any further debugging as there isn't really any visibility once I see the Cookie returned in the successful login request, and am just expecting to be available in Safari as per Apple's documentation.

Happy to provide further information if/as needed.

**Comment by Widcket** [comment]
Hi @chookhen60, thanks for raising this.

> It also works perfectly if the mobile login pathway is replicated (via using the b2c login URL exactly as is presented in the App) in the Safari browser on iOS, and then have the user browse to the website

Could you please provide more details on the scenario in which SSO works? This would be very helpful.

**Comment by chookhen60** [comment]
Hi @Widcket,

SSO works in the following scenarios:
 - Between our Android App (which uses Chrome Custom Tabs as per this library) and the Mobile Google Chrome browser - after manually logging into our App, the user can then go to Mobile Chrome, and navigate to the website (which requires login, so will redirect to the b2c login page when this occurs), has the same auth IdP as the App (b2c), and SSO is achieved, immediately logging them into the website and displaying the the configured page
 - Within the Safari Browser itself (although this is more of a persistence test), by performing the following test:
   - Taking the b2c login URL which the iOS app uses for b2c login, and copy/pasting this into Safari, which gives the user the b2c login screen as they would see within the App (which is displayed in the App within an `AsWebAuthebticationSession` component)
   - Once the user completes their login, it does then try to redirect to the App, but ignoring this part, I switch back to Safari and close the browser
   - Re-open the browser, and navigate to the website which has the same auth IdP, and SSO (or rather persistence) is achieved, immediately logging them into the website and displaying the the configured page.

We currently only have our App and one external website set up, so our scenario testing is limited.  Let me know if this does not answer your questions.

**Comment by Widcket** [comment]
Hi @chookhen60, are you able to reproduce if you previously clear the cookies on Safari?

**Comment by chookhen60** [comment]
Hi @Widcket, yes I can re-produce the issue if I clear the Safari Cookies before starting the whole App login process (I believe this is what you were asking?).  As expected, once the Safari cookies are cleared, trying to access the website in Safari directly will require login as the Cookie is gone.

**Comment by Widcket** [comment]
Sorry, I should have been clearer. I meant to say clear the cookies and then login in the web app, then try SSO on the mobile app:

https://user-images.githubusercontent.com/5055789/158291757-60780724-24ea-402b-88c1-0da6dd5b9a04.MP4

**Comment by chookhen60** [comment]
@Widcket Thanks for clarifying - I'll test this scenario shortly and let you know

**Comment by chookhen60** [comment]
Hi @Widcket, that scenario does not work for me either - after logging into the Website in Safari (and having cleared all Safari data cleared before doing so like in your video) and then switching back to the App and attempting login does not SSO, and the App instead asks me to login (i.e. no SSO achieved).

**Comment by chookhen60** [comment]
@Widcket We aren't using a custom domain as of this moment, but are in the process of setting one up so we can have a more palatable message to users when they sign in.  I'll need to confirm if the App is in the same domain as what we are setting up as the custom domain and let you know.

**Comment by Widcket** [comment]
Please make sure to test on a real device. I dug into this some more, and while it works on devices when the Custom Domain is set up, it does not seem to work on simulators. Could be a bug in the iOS simulator.

**Comment by chookhen60** [comment]
Hi @Widcket - thanks for that info - I had my fears there could be an issue with the Simulator, particularly due to the fact that it sounds like the way Cookies are stored may be different on the iOS simulator. 

I do now have the custom domain working in the app, but still need to configure the website to use this too so both logins are using the same domain.

I have also confirmed our App is in the same domain as the newly added custom domain.

I'm happy to test on a real device when we are able to do so and let you know how that goes.

**Comment by Widcket** [comment]
@chookhen60 are you still experiencing this issue?

**Comment by chookhen60** [comment]
Hi @Widcket, apologies for the delay, we are just trying to get some time to test on a real device, but hoping that will happen today for us. I need to get one of my colleagues to do the testing as I don't have a physical iOS device. 

 I will report back as soon as I have an answer on that testing.

**Comment by chookhen60** [comment]
Hi @Widcket, I can confirm using a real device that the SSO works perfectly using a custom domain for both the app and the website in our case.  Our test case of first logging into the App, then going to Safari and accessing the website, correctly performs SSO on the website and the user does not have to login a second time.

Thank you for your help working through this issue! It's a bit of a shame the Simulator doesn't work for this kind of functionality, but at least the real-world experience for our users will be as we want it to be.

**Comment by Widcket** [comment]
Perfect, thanks for the update. Will close the issue.

**Comment by codycraig** [comment]
I am facing same issue here. When you say custom domain what do you mean exactly? As in a custom URL scheme? https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app @chookhen60

**Comment by chookhen60** [comment]
@codycraig when we say custom domain, we mean not the default domain of the IdP, which in my case is b2clogin.com. We used a custom domain so the IdP base is for my website instead, for example mywebsite.com. This is particularly useful as the SSO message that appear during the login flow references the domain name, and it can't be customised, so using a custom domain, ideally the same one as the website/app operates in, gives a better message to the user.

**Comment by arsencenko** [comment]
@chookhen60 im having the same issue as you with azure b2c, react native msal. I have also tried and tested on real device and it seems that ASWebAuthenticationSession is not shared within Safari, can u please confirm that using a custom domain instead of xxx.b2clogin.com fixed the issue for you or did u changed something else ?

**Comment by chookhen60** [comment]
Hi @arsencenko, It has been a little while for me now, but yes, I can confirm  using a real device plus the custom domain did resolve the issue for me.

From memory (and having a quick look back at this thread) the custom domain alone was not enough to get it to work, but at the very least it will only work on a real device.  We never tested a real device without the custom domain, as it was already in place by that stage. 

I would suggest trying on a real device first if you are not otherwise planning on using a custom domain, and if that isn't working still, I'd go ahead with a custom domain.

**Comment by arsencenko** [comment]
@chookhen60 Thanks for quick response, i have tried on real device iOS 17.2 and its not working the session/cookies are not beign shared from my app to safari. in android it works perfectly. MSAL says that they use ASWebAuthenticationSession as default view and i can see that im also usign prefersEphemeralWebBrowserSession but still no success. maybe custom domain will do the trick. btw did u used react-native-msal ?

**Comment by chookhen60** [comment]
@arsencenko You will need to ensure that `ephemaralSession` is set to false (which you may be, but I wasn't sure from your last response) as that will prevent any sharing from the App to the browser. I am using a version of this libarary with a few small modifications for our purposes here.

**Comment by arsencenko** [comment]
@chookhen60 i will have to try custom domain first and home that works.

**Comment by arsencenko** [comment]
@chookhen60 i have changed azure b2c to use custom domain, im actually seeing the safari data that now is saving some data for my domain but when i open my SAML app url safari still requesting login credentials and redirects to login screen. :( This is on real device

**Comment by chookhen60** [comment]
@arsencenko It's a bit tricky to know at this stage why that may be the case without knowing more about the specifics how how your set up is configured.

It might be worth raising a separate issue at this stage so those with more knowledge than me can assist (and where you can provide the additional detail that is required as part of the issue creation).

**Comment by arsencenko** [comment]
@chookhen60 what react native library are you using to achieve azure b2c login ?

**Comment by chookhen60** [comment]
@arsencenko as mentioned above, we are using a forked version of this libary `react-native-auth0` which some changes for us to perform a successful b2c login.  This fork pre-dates my time on the project this is used in, so this is all the information I know, and I don't know the exact details of what changes were needed, although I don't believe they were overly large

**Comment by arsencenko** [comment]
@chookhen60 im using react-native-msal maybe thats where the issue stands. but still im not sure. Does your up work at the moment with this flow ?
